finrl/env/env_stocktrading_cashpenalty.py

```diff
@@ -296,68 +268,64 @@ def step(self, actions):
             else:
                 actions = np.where(closings > 0, actions / closings, 0)
 
-            # clip actions so we can't sell more assets than we hold
-            actions = np.maximum(actions, -np.array(holdings))
-
             # round down actions to the nearest multiplies of shares_increment
-            actions = (actions // self.shares_increment) * self.shares_increment
-
-            self.transaction_memory.append(actions)  # capture what the model's could do
+            actions = np.where(actions >= 0,
+                            (actions // self.shares_increment) * self.shares_increment,
+                            ((actions + self.shares_increment) // self.shares_increment) * self.shares_increment)
 
+            # clip actions so we can't sell more assets than we hold
+            actions = np.maximum(actions, -np.array(holdings))
             # compute our proceeds from sells, and add to cash
             sells = -np.clip(actions, -np.inf, 0)
             proceeds = np.dot(sells, closings)
             costs = proceeds * self.sell_cost_pct
             coh = begin_cash + proceeds
-
             # compute the cost of our buys
             buys = np.clip(actions, 0, np.inf)
             spend = np.dot(buys, closings)
             costs += spend * self.buy_cost_pct
-
-            # if we run out of cash, end the cycle and penalize
+             # if we run out of cash...
             if (spend + costs) > coh:
-                return self.return_terminal(
-                    reason="CASH SHORTAGE", reward=self.get_reward()
-                )
-
+                if self.patient:
+                    # ... just don't buy anything until we got additional cash
+                    self.log_step(reason="CASH SHORTAGE")
+                    actions = np.where(actions>0,0,actions)
+                    spend = 0
+                    costs = 0
+                else:
+                    # ... end the cycle and penalize
+                    return self.return_terminal(
+                        reason="CASH SHORTAGE",reward=self.get_reward()
+                    )
+            self.transaction_memory.append(actions)  # capture what the model's could do
             # verify we didn't do anything impossible here
```

Comment:
Can we please remove this feature from this PR and plan to do it in a subsequent PR?

Reply:
Sure, will create other PR for adding the `self.patient` functionality.

Reply:
Also adding a test for share_increment later. Thanks!